odl/space/fspace.py

```diff
@@ -627,88 +308,310 @@ def element(self, fcall=None, vectorized=True):
             return fcall
-        else:
-            if not callable(fcall):
-                raise TypeError('`fcall` {!r} is not callable'.format(fcall))
+        elif callable(fcall):
             if not vectorized:
+                if hasattr(fcall, 'nin') and hasattr(fcall, 'nout'):
+                    raise TypeError('`fcall` {!r} is a ufunc-like object, '
+                                    'use vectorized=True')
+                has_out, _ = _check_out_arg(fcall)
+                if has_out:
+                    raise TypeError('non-vectorized `fcall` with `out` '
+                                    'parameter not allowed')
                 if self.field == RealNumbers():
-                    dtype = 'float64'
+                    otypes = ['float64']
+                elif self.field == ComplexNumbers():
+                    otypes = ['complex128']
                 else:
-                    dtype = 'complex128'
-
-                fcall = vectorize(otypes=[dtype])(fcall)
+                    otypes = []
 
+                fcall = vectorize(otypes=otypes)(fcall)
             return self.element_type(self, fcall)
+        else:
+            # This is for the case that an array-like of callables
+            # is provided
+            if np.shape(fcall) != self.out_shape:
+                raise ValueError(
+                    'invalid `fcall` {!r}: expected `None`, a callable or '
+                    'an array-like of callables whose shape matches '
+                    '`out_shape` {}'.format(self.out_shape))
+
+            fcalls = np.array(fcall, dtype=object, ndmin=1).ravel().tolist()
+            if not vectorized:
+                if self.field == RealNumbers():
+                    otypes = ['float64']
+                elif self.field == ComplexNumbers():
+                    otypes = ['complex128']
+                else:
+                    otypes = []
 
-    def zero(self):
-        """Function mapping everything to zero.
+                # Vectorize, preserving scalars
+                fcalls = [f if np.isscalar(f) else vectorize(otypes=otypes)(f)
+                          for f in fcalls]
 
-        This function is the additive unit in the function space.
+            if sys.version_info.major < 3:
+                getargspec = inspect.getargspec
+            else:
+                getargspec = inspect.getfullargspec
+
+            def wrapper(x, out=None, **kwargs):
+                """Function wrapping an array of callables."""
+                if is_valid_input_meshgrid(x, self.domain.ndim):
+                    scalar_out_shape = out_shape_from_meshgrid(x)
+                elif is_valid_input_array(x, self.domain.ndim):
+                    scalar_out_shape = out_shape_from_array(x)
+                else:
+                    raise RuntimeError('bad input')
+
+                if out is None:
+                    # Out-of-place evaluation
+
+                    # Collect results of member functions into a list
```

Comment:
I don't think I can make it simpler or more readable since there are really a bunch of nasty corner cases to be handled. But I'll add a more explanatory docstring to the wrapper.